distributed/worker.py

```diff
@@ -2853,19 +2885,27 @@ def _put_key_in_memory(self, ts: TaskState, value, *, stimulus_id: str) -> Recs:
         self.log.append((ts.key, "put-in-memory", stimulus_id, time()))
         return recommendations
 
-    def select_keys_for_gather(self, worker, dep):
-        assert isinstance(dep, str)
+    def _select_keys_for_gather(
+        self, worker: str, dep: str, all_keys_to_gather: Container[str]
+    ) -> tuple[set[str], int]:
+        """``_ensure_communicating`` decided to fetch a single task from a worker,
+        following priority. In order to minimise overhead, request fetching other tasks
+        from the same worker within the message, following priority for the single
+        worker but ignoring higher priority tasks from other workers, up to
+        ``target_message_size``.
+        """
         deps = {dep}
-
         total_bytes = self.tasks[dep].get_nbytes()
-        L = self.pending_data_per_worker[worker]
+        pending = self.pending_data_per_worker[worker]
 
-        while L:
-            ts = L.pop()
-            if ts.state != "fetch":
+        while pending:
+            ts = pending.peek()
```

Comment:
Fix bug where a task that would cause ``total_bytes`` to exceed ``target_message_size`` was excluded from future iterations of ``select_keys_for_gather``